modules/sw360/sw360-workflow/src/main/java/org/eclipse/sw360/antenna/sw360/SW360MetaDataUpdater.java

```diff
@@ -108,15 +112,74 @@ public boolean isUploadSources() {
-        LOGGER.debug("Result of attachment upload operation: {}", result);
-        if (!result.isSuccess()) {
-            LOGGER.error("Failed to upload attachments: {}", result.failedUploads());
+        AttachmentUploadRequest<SW360Release> uploadRequest = builder.build();
+        if (!uploadRequest.getItems().isEmpty()) {
+            AttachmentUploadResult<SW360Release> result = releaseClientAdapter.uploadAttachments(uploadRequest);
+            LOGGER.debug("Result of attachment upload operation: {}", result);
+            if (!result.isSuccess()) {
+                LOGGER.error("Failed to upload attachments: {}", result.failedUploads());
+            }
+            return result.getTarget();
+        }
+        return sw360Release;
+    }
+
+    /**
+     * Calculates the SHA-1 hash for the given local attachment file. This is
+     * used to determine whether a modified attachment file needs to be
+     * uploaded.
+     *
+     * @param path the path to the local file
+     * @return the SHA-1 hash of this file
+     * @throws SW360ClientException if an error occurs
+     */
+    String calculateAttachmentHash(Path path) {
+        return SW360AttachmentUtils.calculateSha1Hash(path);
+    }
+
+    /**
+     * Checks whether an attachment upload is necessary for the given
+     * attachment file. This is the case if no attachment with this file name
+     * and type and identical content exists.
+     *
+     * @param release        the release affected
+     * @param path           the path to the local attachment file
+     * @param attachmentType the attachment type
+     * @return a flag whether this attachment file must be uploaded
+     */
+    private boolean needUpload(SW360Release release, Path path, SW360AttachmentType attachmentType) {
+        String fileName = String.valueOf(path.getFileName());
+        Optional<SW360SparseAttachment> optAttachment = release.getEmbedded().getAttachments().stream()
+                .filter(attachment -> attachmentMatches(path, attachmentType, fileName, attachment))
+                .findFirst();
+        return !optAttachment.isPresent();
+    }
+
+    /**
+     * Checks whether an attachment associated with a release matches the given
+     * properties. If there is an exact match, the upload of this attachment
+     * file can be skipped.
+     *
+     * @param path           the path to the local attachment file
+     * @param attachmentType the attachment type
+     * @param fileName       the attachment file name
+     * @param attachment     the attachment from the current release
+     * @return a flag whether the attachment is matched by these properties
+     */
+    private boolean attachmentMatches(Path path, SW360AttachmentType attachmentType, String fileName, SW360SparseAttachment attachment) {
+        try {
+            return Objects.equals(fileName, attachment.getFilename()) &&
+                    attachmentType == attachment.getAttachmentType() &&
```

Comment:
if they have a different type, but the same filename and sha1 they shouldn't be uploaded, since they can't be of different types if it is the same file